### PR TITLE
[native] Adapt new async cache config

### PR DESCRIPTION
**Pull request by zacw7**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by amitkdutta — COMMENTED** [review]
A small nit. Looks great otherwise.

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 1]
```diff
@@ -320,6 +320,10 @@ class SystemConfig : public ConfigBase {
   /// option to disable cow for cache files.
   static constexpr std::string_view kAsyncCacheSsdDisableFileCow{
       "async-cache-ssd-disable-file-cow"};
+  static constexpr std::string_view kAsyncCacheSsdChecksumEnabled{
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:323]:
```
s/kAsyncCacheSsdChecksumEnabled/kSsdCacheChecksumEnabled/
s/async-cache-ssd-checksum-enabled/ssd-cache-checksum-enabled/
```

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 2]
```diff
@@ -320,6 +320,10 @@ class SystemConfig : public ConfigBase {
...
+      "async-cache-ssd-checksum-enabled"};
+  static constexpr std::string_view kAsyncCacheSsdReadVerificationEnabled{
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:325]:
```
s/kAsyncCacheSsdReadVerificationEnabled/kSsdCacheReadVerificationEnabled/
s/async-cache-ssd-read-verification-enabled/ssd-cache-read-verification-enabled/
```

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 3]
```diff
@@ -625,6 +629,10 @@ class SystemConfig : public ConfigBase {
 
   bool asyncCacheSsdDisableFileCow() const;
 
+  bool asyncCacheSsdChecksumEnabled() const;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:632]: ditto

File: presto-native-execution/presto_cpp/main/common/Configs.h (hunk 1 above)
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:323]: Can you briefly comment these configs? thanks!

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 4]
```diff
@@ -701,22 +701,32 @@ void PrestoServer::initializeVeloxMemory() {
...
-          systemConfig->asyncCacheSsdDisableFileCow();
+      cache::Options cacheOptions;
+      cacheOptions.checkpointIntervalBytes = asyncCacheSsdCheckpointGb << 30;
+      cacheOptions.checksumEnabled =
+          systemConfig->asyncCacheSsdChecksumEnabled();
+      cacheOptions.checksumReadVerificationEnabled =
+          systemConfig->asyncCacheSsdReadVerificationEnabled();
+      cacheOptions.disableFileCow = systemConfig->asyncCacheSsdDisableFileCow();
       PRESTO_STARTUP_LOG(INFO)
           << "Initializing SSD cache with capacity " << asyncCacheSsdGb
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:714]: Cache::Options::toString() to logging.

**Review by xiaoxmeng — COMMENTED** [review]
@zacw7 thanks for the change!

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 5]
```diff
@@ -701,22 +701,32 @@ void PrestoServer::initializeVeloxMemory() {
...
+                                                           : "DISABLED");
+
       ssd = std::make_unique<cache::SsdCache>(
           systemConfig->asyncCacheSsdPath(),
           asyncCacheSsdGb << 30,
           kNumSsdShards,
-          cacheExecutor_.get(),
-          asyncCacheSsdCheckpointGb << 30,
-          asyncCacheSsdDisableFileCow);
+          cacheOptions,
```
[the remark this answers is not in the document]
Comment (zacw7) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp]: it's a bit tricky as there is a member `cacheExecutor_` holding a unique ptr.

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 6]
```diff
@@ -320,6 +320,15 @@ class SystemConfig : public ConfigBase {
   /// option to disable cow for cache files.
   static constexpr std::string_view kAsyncCacheSsdDisableFileCow{
       "async-cache-ssd-disable-file-cow"};
+  /// When enabled, a CRC-based checksum is calculated for each cache entry and
+  /// stored in the checkpoint file.
+  static constexpr std::string_view kSsdCacheChecksumEnabled{
+      "ssd-cache-checksum-enabled"};
+  /// When enabled, the checksum is recalculated and verified against the stored
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:327]:
```
  /// When enabled, the checksum is recalculated and verified against the stored value on SSD when load data from SSD.
```
I don't think we do checksum verification during checkpoint recover?

File: presto-native-execution/presto_cpp/main/common/Configs.h [hunk 7]
```diff
@@ -320,6 +320,15 @@ class SystemConfig : public ConfigBase {
   /// option to disable cow for cache files.
   static constexpr std::string_view kAsyncCacheSsdDisableFileCow{
       "async-cache-ssd-disable-file-cow"};
+  /// When enabled, a CRC-based checksum is calculated for each cache entry and
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/common/Configs.h:323]:
```
  /// When enabled, a CRC-based checksum is calculated for each cache entry written to SSD. The checksum is stored in the next checkpoint file.
```

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 8]
```diff
@@ -701,22 +701,18 @@ void PrestoServer::initializeVeloxMemory() {
...
-          systemConfig->asyncCacheSsdCheckpointGb();
-      auto asyncCacheSsdDisableFileCow =
-          systemConfig->asyncCacheSsdDisableFileCow();
-      PRESTO_STARTUP_LOG(INFO)
-          << "Initializing SSD cache with capacity " << asyncCacheSsdGb
-          << "GB, checkpoint size " << asyncCacheSsdCheckpointGb
-          << "GB, file cow "
-          << (asyncCacheSsdDisableFileCow ? "DISABLED" : "ENABLED");
-      ssd = std::make_unique<cache::SsdCache>(
+      cache::SsdCache::Config cacheConfig(
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:704]: Let's have an anonymous function say
```
const auto ssdCacheConfig = getSsdCacheConfig()?
```

**Review by xiaoxmeng — COMMENTED** [review]
@zacw7 thanks for the update!

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 9]
```diff
@@ -701,22 +714,10 @@ void PrestoServer::initializeVeloxMemory() {
       constexpr int32_t kNumSsdShards = 16;
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:714]: How about we just capture everything in
```
std::unique_ptr<cache::SsdCache> ssd = setupSsdCache();
```

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 10]
```diff
@@ -696,27 +714,8 @@ void PrestoServer::initializeVeloxMemory() {
...
-          << "GB, file cow "
-          << (asyncCacheSsdDisableFileCow ? "DISABLED" : "ENABLED");
-      ssd = std::make_unique<cache::SsdCache>(
-          systemConfig->asyncCacheSsdPath(),
-          asyncCacheSsdGb << 30,
-          kNumSsdShards,
-          cacheExecutor_.get(),
-          asyncCacheSsdCheckpointGb << 30,
-          asyncCacheSsdDisableFileCow);
+    if (systemConfig->asyncCacheSsdGb() > 0) {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:717]: Can we move this check into setupSsdCache?

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 11]
```diff
@@ -651,6 +651,24 @@ void PrestoServer::initializeThreadPools() {
   }
 }
 
+std::unique_ptr<cache::SsdCache> PrestoServer::setupSsdCache() {
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:654]:
```
VELOX_CHECK_NULL(cacheExecutor_);
```

**Review by xiaoxmeng — DISMISSED** [review]
@zacw7 thanks for the update!

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp (hunk 11 above)
Comment (zacw7) [reply presto-native-execution/presto_cpp/main/PrestoServer.cpp:654]: Is this needed right after we created cacheExecutor_? Also we are already checking it inside SsdCache.

File: presto-native-execution/presto_cpp/main/PrestoServer.cpp [hunk 12]
```diff
@@ -696,28 +718,7 @@ void PrestoServer::initializeVeloxMemory() {
...
-          << (asyncCacheSsdDisableFileCow ? "DISABLED" : "ENABLED");
-      ssd = std::make_unique<cache::SsdCache>(
-          systemConfig->asyncCacheSsdPath(),
-          asyncCacheSsdGb << 30,
-          kNumSsdShards,
-          cacheExecutor_.get(),
-          asyncCacheSsdCheckpointGb << 30,
-          asyncCacheSsdDisableFileCow);
-    }
+    ssd = setupSsdCache();
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.cpp:721]:
```
std::unique_ptr<cache::SsdCache> ssd = setupSsdCache();
```

File: presto-native-execution/presto_cpp/main/PrestoServer.h [hunk 13]
```diff
@@ -256,6 +256,9 @@ class PrestoServer {
   std::string address_;
   std::string nodeLocation_;
   folly::SSLContextPtr sslContext_;
+
+ private:
```
Comment (xiaoxmeng) [inline presto-native-execution/presto_cpp/main/PrestoServer.h:260]: Move up this function declaration.

**Review by xiaoxmeng — APPROVED** [review]

**Pull request merged**